File: 2017/day24/day24.rs
//! [Day 24: Electromagnetic Moat](https://adventofcode.com/2017/day/24)

use std::collections::VecDeque;

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (u32, u32) {
    let ports: Vec<_> = data
        .lines()
        .filter_map(|line| line.split_once('/'))
        .map(|(a, b)| (a.parse().unwrap(), b.parse().unwrap()))
        .collect();

    let mut max_length = 0;
    let mut queue = VecDeque::new();

    let mut max_strength = 0; // part 1 answer
    let mut max_length_strength = 0; // part 2 answer

    queue.push_front((0u32, 0, 1, ports));
    while let Some((pin, strength, length, ports)) = queue.pop_back() {
        //
        max_strength = max_strength.max(strength);

        // match expression is equivalent to the following, clippy is sometimes really weird
        // if length > max_length {
        //     max_length = length;
        //     max_length_strength = strength;
        // } else if length == max_length {
        //     max_length_strength = max_length_strength.max(strength);
        // }

        match length {
            _ if length > max_length => {
                max_length = length;
                max_length_strength = strength;
            }
            _ if length == max_length => {
                max_length_strength = max_length_strength.max(strength);
            }
            _ => {}
        };

        for (i, &(a, b)) in ports.iter().enumerate() {
            let c = if a == pin {
                b
            } else if b == pin {
                a
            } else {
                continue;
            };

            let mut np = ports.clone();
            np.remove(i);

            queue.push_front((c, strength + a + b, length + 1, np));
        }
    }

    (max_strength, max_length_strength)
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test_solve() {
        let answer = solve(TEST_INPUT);
        assert_eq!(answer.0, 31);
        assert_eq!(answer.1, 19);
    }
}

File: src/year2017/day24/day24.py
#!/usr/bin/env python3
# [Day 24: Electromagnetic Moat](https://adventofcode.com/2017/day/24)

import atexit
import sys
import time
from collections import deque
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
self_tests = "-T" in sys.argv
if self_tests:
    sys.argv.remove("-T")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


ports = [tuple(map(int, line.split("/"))) for line in data.splitlines()]


max_strength = 0
max_length = 0
max_length_strength = 0
q = deque()
q.append((0, 0, 1, ports))
while q:
    pin, strength, length, ports = q.popleft()

    max_strength = max(max_strength, strength)

    if length > max_length:
        max_length = length
        max_length_strength = strength
    elif length == max_length:
        max_length_strength = max(max_length_strength, strength)

    for i, (a, b) in enumerate(ports):
        if a == pin:
            c = b
        elif b == pin:
            c = a
        else:
            c = None
        if c is not None:
            np = list(ports)
            np.pop(i)
            q.append((c, strength + a + b, length + 1, np))

# part 1
print(max_strength)

# part 2
print(max_length_strength)

# tests
if filename == "test.txt":
    assert max_strength == 31
    assert max_length_strength == 19
